Migrate to AWS SDK v2

anas-aso opened the pull request:
Signed-off-by: Anas anas.oubrahim@commercetools.com

Fixes https://github.com/prometheus/cloudwatch_exporter/issues/172

@brian-brazil can I have your review please ?

brian-brazil commented on pom.xml:
```diff
@@ -41,22 +41,32 @@
   <dependencies>
 
     <dependency>
-      <groupId>com.amazonaws</groupId>
```
Please leave the dependency style as it was

anas-aso replied:
Done : https://github.com/prometheus/cloudwatch_exporter/pull/307/commits/694c71e035c532588d534dacb4f4c5fe9c3a2c4e

brian-brazil commented on src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:
```diff
@@ -1,54 +1,42 @@
-import java.util.Map;
-import java.util.HashMap;
-import java.util.HashSet;
+import java.util.*;
```
Please keep imports explicit.

anas-aso replied:
Done https://github.com/prometheus/cloudwatch_exporter/pull/307/commits/8feacfb8405cd99c30f0d2650688d58e5bc47559

brian-brazil commented on src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:
```diff
@@ -214,9 +202,11 @@ private void loadConfig(Map<String, Object> config, AmazonCloudWatch cloudWatchC
           if (yamlMetricRule.containsKey("aws_statistics")) {
-            rule.awsStatistics = (List<String>)yamlMetricRule.get("aws_statistics");
+            rule.awsStatistics = ((List<String>)yamlMetricRule.get("aws_statistics"))
+                    .stream().map(e -> Statistic.fromValue(e)).collect(Collectors.toList());
```
Please use a for loop here, we're not using the newer syntaxes in this repo.

anas-aso replied:
Done: https://github.com/prometheus/cloudwatch_exporter/pull/307/commits/24b5f95ce54ced965ec7d2b17b12c311f77e0d22

brian-brazil reviewed (commented):
From a cursory review this looks okay, however I expect the devil is in the details. What sort of manual testing have you done on this?

anas-aso commented:
@brian-brazil I have addressed your reviews. Can you please have a 2nd look ?

> From a cursory review this looks okay, however I expect the devil is in the details. What sort of manual testing have you done on this?

I run the latest released version and this PR locally, fetch the metrics (wget) and compare the output (I did it multiple times). The result is always like this : 
```
2a3,6
> # HELP cloudwatch_requests_total API requests made to CloudWatch
> # TYPE cloudwatch_requests_total counter
> cloudwatch_requests_total{action="getMetricStatistics",namespace="AWS/SQS",} 5117.0
> cloudwatch_requests_total{action="listMetrics",namespace="AWS/SQS",} 119.0
637c641
< cloudwatch_exporter_scrape_duration_seconds 25.683923456
---
> cloudwatch_exporter_scrape_duration_seconds 25.514181182
641,644d644
< # HELP cloudwatch_requests_total API requests made to CloudWatch
< # TYPE cloudwatch_requests_total counter
< cloudwatch_requests_total{action="getMetricStatistics",namespace="AWS/SQS",} 27692.0
< cloudwatch_requests_total{action="listMetrics",namespace="AWS/SQS",} 644.0
```

I am also running a local Prometheus and it's scrapping both version of the exporter : 
<img width="1360" alt="count" src="https://user-images.githubusercontent.com/16538065/98267687-efdbb280-1f8b-11eb-90ee-c008aa28e0d6.png">
<img width="1360" alt="sum" src="https://user-images.githubusercontent.com/16538065/98267692-f23e0c80-1f8b-11eb-8bca-e6f5e33d3c0d.png">

Not perfect, but should give a rough idea about the returned results by both versions of the exporter.

PS: I am importing SQS metrics only.

If you have a better idea for manual testing I can give it a try (if it doesn't require too much time).

brian-brazil commented on src/main/java/io/prometheus/cloudwatch/CloudWatchCollector.java:
```diff
@@ -42,13 +22,28 @@
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.cloudwatch.CloudWatchClient;
+import software.amazon.awssdk.services.cloudwatch.CloudWatchClientBuilder;
+import software.amazon.awssdk.services.cloudwatch.model.*;
```
This one isn't explicit.

anas-aso replied:
Done : https://github.com/prometheus/cloudwatch_exporter/pull/307/commits/c7fd02fba98cc0bca1cc30e13510be148c647388

brian-brazil reviewed (commented):
Try all the config options basically, the tag ones in particular as that's a separate code path. Something is likely to break with a change like this, but we should try to minimise the chances of it.

anas-aso commented:
@brian-brazil I have tried all the config options and all seems fine. It was time consuming but I think it was worth it since I found something that I missed before https://github.com/prometheus/cloudwatch_exporter/pull/307/commits/ecd822ffbe49bbf2ffc96d8fc274ff044b8f3640

Regarding the tagging feature, here is the results of the tests I did:
#### Test 1: Without tag filtering
Using the config below:
```yaml
region: eu-central-1
metrics:
  - aws_namespace: AWS/ApplicationELB
    aws_metric_name: RequestCount
    aws_dimensions: [LoadBalancer]
    aws_statistics: [Sum]
```
I got the following results
```
# HELP cloudwatch_requests_total API requests made to CloudWatch
# TYPE cloudwatch_requests_total counter
cloudwatch_requests_total{action="listMetrics",namespace="AWS/ApplicationELB",} 4.0
cloudwatch_requests_total{action="getMetricStatistics",namespace="AWS/ApplicationELB",} 33.0
# HELP tagging_api_requests_total API requests made to the Resource Groups Tagging API
# TYPE tagging_api_requests_total counter
# HELP aws_applicationelb_request_count_sum CloudWatch metric AWS/ApplicationELB RequestCount Dimensions: [LoadBalancer] Statistic: Sum Unit: Count
# TYPE aws_applicationelb_request_count_sum gauge
aws_applicationelb_request_count_sum{job="aws_applicationelb",instance="",load_balancer="app/REDACTED",} 14.0 1604676480000
aws_applicationelb_request_count_sum{job="aws_applicationelb",instance="",load_balancer="app/REDACTED",} 6.0 1604676480000
aws_applicationelb_request_count_sum{job="aws_applicationelb",instance="",load_balancer="app/REDACTED",} 2.0 1604676480000
aws_applicationelb_request_count_sum{job="aws_applicationelb",instance="",load_balancer="app/REDACTED",} 4.0 1604676480000
aws_applicationelb_request_count_sum{job="aws_applicationelb",instance="",load_balancer="app/REDACTED",} 3.0 1604676480000
aws_applicationelb_request_count_sum{job="aws_applicationelb",instance="",load_balancer="app/REDACTED",} 4.0 1604676480000
aws_applicationelb_request_count_sum{job="aws_applicationelb",instance="",load_balancer="app/REDACTED",} 4.0 1604676480000
aws_applicationelb_request_count_sum{job="aws_applicationelb",instance="",load_balancer="app/REDACTED",} 0.0 1604676480000
aws_applicationelb_request_count_sum{job="aws_applicationelb",instance="",load_balancer="app/REDACTED",} 2.0 1604676480000
aws_applicationelb_request_count_sum{job="aws_applicationelb",instance="",load_balancer="app/REDACTED",} 2.0 1604676480000
# HELP aws_resource_info AWS information available for resource
# TYPE aws_resource_info gauge
# HELP cloudwatch_exporter_scrape_duration_seconds Time this CloudWatch scrape took, in seconds.
# TYPE cloudwatch_exporter_scrape_duration_seconds gauge
cloudwatch_exporter_scrape_duration_seconds 0.377298838
# HELP cloudwatch_exporter_scrape_error Non-zero if this scrape failed.
# TYPE cloudwatch_exporter_scrape_error gauge
cloudwatch_exporter_scrape_error 0.0
```

#### Test 2: With tag filtering
Using the config below:
```yaml
region: eu-central-1
metrics:
  - aws_namespace: AWS/ApplicationELB
    aws_metric_name: RequestCount
    aws_dimensions: [LoadBalancer]
    aws_statistics: [Sum]
    aws_tag_select:
      tag_selections:
        kubernetes.io/ingress-name: ["REDACTED"]
      resource_type_selection: "elasticloadbalancing:loadbalancer"
      resource_id_dimension: LoadBalancer
```
I got the following results
```
# HELP cloudwatch_requests_total API requests made to CloudWatch
# TYPE cloudwatch_requests_total counter
cloudwatch_requests_total{action="listMetrics",namespace="AWS/ApplicationELB",} 10.0
cloudwatch_requests_total{action="getMetricStatistics",namespace="AWS/ApplicationELB",} 64.0
# HELP tagging_api_requests_total API requests made to the Resource Groups Tagging API
# TYPE tagging_api_requests_total counter
tagging_api_requests_total{action="getResources",resource_type="elasticloadbalancing:loadbalancer",} 4.0
# HELP aws_applicationelb_request_count_sum CloudWatch metric AWS/ApplicationELB RequestCount Dimensions: [LoadBalancer] Statistic: Sum Unit: Count
# TYPE aws_applicationelb_request_count_sum gauge
aws_applicationelb_request_count_sum{job="aws_applicationelb",instance="",load_balancer="app/REDACTED",} 6.0 1604676480000
# HELP aws_resource_info AWS information available for resource
# TYPE aws_resource_info gauge
aws_resource_info{job="aws_applicationelb",instance="",arn="arn:aws:elasticloadbalancing:eu-central-1:REDACTED:loadbalancer/app/REDACTED",load_balancer="app/REDACTED",tag_ingress_k8s_aws_cluster="REDACTED",tag_ingress_k8s_aws_resource="LoadBalancer",tag_ingress_k8s_aws_stack="REDACTED",tag_kubernetes_io_cluster_REDACTED="REDACTED",tag_kubernetes_io_ingress_name="REDACTED",tag_kubernetes_io_namespace="REDACTED",} 1.0
# HELP cloudwatch_exporter_scrape_duration_seconds Time this CloudWatch scrape took, in seconds.
# TYPE cloudwatch_exporter_scrape_duration_seconds gauge
cloudwatch_exporter_scrape_duration_seconds 0.138822136
# HELP cloudwatch_exporter_scrape_error Non-zero if this scrape failed.
# TYPE cloudwatch_exporter_scrape_error gauge
cloudwatch_exporter_scrape_error 0.0
```

anas-aso commented:
@brian-brazil is there anything else needed for this PR ?

brian-brazil commented:
I haven't had a chance to test this out myself yet.

brian-brazil commented:
Thanks!
